worker/controlsocket/handlers.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controlsocket
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/environs/bootstrap"
+	"github.com/juju/juju/state"
+	stateerrors "github.com/juju/juju/state/errors"
+)
+
+const (
+	// jujuMetricsUserPrefix defines the "namespace" in which this worker is
+	// allowed to create/remove users.
+	jujuMetricsUserPrefix = "juju-metrics-"
+
+	// userCreator is the listed "creator" of metrics users in state.
+	// This user CANNOT be a local user (it must have a domain), otherwise the
+	// model addUser code will complain about the user not existing.
+	userCreator = "controller@juju"
+)
+
+func (w *Worker) registerHandlers(r *mux.Router) {
+	// We're using a fully REST-ful API structure here.
```

Comment:
Bit of trivia; _Fully_ REST-ful is technically impossible when you use JSON, if you base the definition on the Richardson Maturity Model.

Reply:
this is true, without the ability to represent hypermedia controls you can't really get to level 3. But, we could adopt https://jsonapi.org/ which would give us that. Not really needed here, level 2 is enough.